lib/foreman_tasks_core/runner/action.rb

```diff
@@ -52,15 +54,34 @@ def finish_run(update)
         output[:exit_status] = update.exit_status
       end
 
+      def process_external_event(event)
+        runner_dispatcher.external_event(output[:runner_id], event)
+        suspend
+      end
+
       def process_update(update)
-        output[:result].concat(update.continuous_output.raw_outputs)
+        merge_outputs(update.continuous_output.raw_outputs, update.incremental)
         if update.exit_status
           finish_run(update)
         else
           suspend
         end
       end
 
+      def merge_outputs(outputs, incremental)
+        # No need to do anything if there are no inputs or if the "new" outputs are older than the already saved ones
+        return if outputs.empty? || (!output[:result].empty? && outputs.first['timestamp'] < output[:result].last['timestamp'])
+        unless incremental
+          # The new output is complete, we need to remove the already known outputs from it
+          # Also since it's complete output, there's only one
+          new = outputs.first
+          known = output[:result].select { |out| out['timestamp'] < outputs.first['timestamp'] }
```

Comment:
Couldn't we just iterate though the known and `gsub!(/\A#{out['output]}/)` iteratively, not needed the timestamp at all? I'm wondering, if we can rely on `timestamp`'s uniqueness and if we there are some other risks around that. If we didn't, I would feel a bit better about that.